### PR TITLE
Add telemetry events

### LINKED ISSUE #68: Telemetry events around http requests
Hey,

we've been using this library for some time now and now we came across a need for putting some metrics around HTTP calls our system makes. We've been thinking of adding some telemetry events emitting to Plaid library itself to allow anybody using it handle them and later report. Our suggestions would be to have a set of events similar to this:
* plaid.request.start (measurement: `system_time`)
* plaid.request.stop (measurement: `duration`)
* plaid.request.exception (measurement: `duration`)
All of them would have metadata containing basic request info (path, status code, url etc.).

We're happy to implement it ourselves, let us know what you think.

### PR DESCRIPTION
**Pull request by kanes115**
Fixes #68 
This PR adds events around HTTP requests made by this library.

### CODE REVIEW

File: test/lib/plaid_test.exs [hunk 1]
```diff
@@ -154,6 +154,91 @@ defmodule PlaidTest do
...
+
+      assert {@start_event, %{system_time: _}, meta} = start
+      assert {@stop_event, %{duration: _}, ^meta} = stop
+
+      assert %{method: :get, path: "any", u: :native} = meta
+    end
+
+    test "are sent when there's an exception", %{bypass: bypass} do
+      # A character outside of utf-8 so that Poison raises
+      body = %{key: <<128>>}
```
Comment (kanes115) [inline test/lib/plaid_test.exs:210]: That's obviously not the best way to test that. The best would be to use Mox, have HTTP layer replacable but that would require to change all other tests. For know I think it's enough to use this little trick and when/if we introduce Mox we can change this place as well.

File: lib/plaid.ex [hunk 2]
```diff
@@ -88,33 +88,39 @@ defmodule Plaid do
...
 
-    try do
-      request_endpoint = "#{get_root_uri(config)}#{endpoint}"
-      cred = Map.delete(config, :root_uri)
-      request_body = Map.merge(body, cred) |> Poison.encode!()
-      request_headers = get_request_headers() |> Map.merge(headers) |> Map.to_list()
-      options = httpoison_request_options() ++ options
+    with_metrics(
+      fn ->
+        request_endpoint = "#{get_root_uri(config)}#{endpoint}"
```
Comment (wfgilman) [inline lib/plaid.ex:99]: What are your thoughts on including the preparation of the request arguments in the instrumentation? I would expect just `request/6` to be instrumented.

File: lib/plaid.ex [hunk 3]
```diff
@@ -88,33 +88,39 @@ defmodule Plaid do
...
-    })
+    }
 
-    try do
-      request_endpoint = "#{get_root_uri(config)}#{endpoint}"
-      cred = Map.delete(config, :root_uri)
-      request_body = Map.merge(body, cred) |> Poison.encode!()
-      request_headers = get_request_headers() |> Map.merge(headers) |> Map.to_list()
-      options = httpoison_request_options() ++ options
+    with_metrics(
```
Comment (wfgilman) [inline lib/plaid.ex:97]: One of the features of instrumentation I had in mind for this library was optionality. Generally I would expect instrumentation to be the responsibility of the application, which might measure Plaid requests alongside other events, such as database calls, using something similar to Appsignal:
```
def make_http_request do
 Appsignal.instrument("plaid-request", fn ->
  Plaid.Transactions.get(...)
 end)
end
```
That said, I think the convenience of out-of-the-box instrumentation for this library is nice feature. I can imagine a scenario where the user has `:telemetry` elsewhere as a dependency in the application, in which case this library could leverage it in `make_request_with_cred/6` in the following way:
```
def make_request_with_cred(method, endpoint, config, body \\ %{}, headers \\ %{}, options \\ []) do
    common_metadata = %{
      method: method,
      path: endpoint,
      u: :native
    }

    request_endpoint = "#{get_root_uri(config)}#{endpoint}"
    cred = Map.delete(config, :root_uri)
    request_body = Map.merge(body, cred) |> Poison.encode!()
    request_headers = get_request_headers() |> Map.merge(headers) |> Map.to_list()
    options = httpoison_request_options() ++ options

    request = fn -> request(method, request_endpoint, request_body, request_headers, options) end

    case Code.ensure_compiled(:telemetry) do
      {:module, _} -> with_metrics(request, common_metadata)
      {:error, _} -> request
    end
  end
```
By making instrumentation optional, users wanting the library just for bindings can keep it lightweight, while those wanting the convenience of standardized instrumentation can take advantage of your feature.

Comment (wfgilman) [reply lib/plaid.ex:97]: Thanks for your thoughts. I'm okay with this implementation given the limitations of the current testing suite.

File: lib/plaid.ex (hunk 2 above)
Comment (wfgilman) [reply lib/plaid.ex:99]: Sounds good to keep as-is given the current limitations on testing. No obligation, but I'd be interested in seeing any examples you might have of Mox and how you handle this issue else.where

Comment (wfgilman) [reply lib/plaid.ex:99]: Thanks, I'd like to explore implementing this. Would you be open to providing a review and some feedback of my implementation?

Also, please let me know when you'd like to turn this draft into a PR and I'll get it merged.

**Comment by kanes115** [comment]
@wfgilman Please rereview and feel free to merge if you approve

**Comment by kanes115** [comment]
@wfgilman just a kind reminder this is ready to be looked at now :)

**Review by wfgilman — APPROVED** [review]

**Pull request merged**